conn.go

```diff
@@ -631,155 +543,202 @@ func (c *Conn) fragmentHandshake(h *handshake) ([][]byte, error) {
 	if err != nil {
-		return &alert{alertLevelFatal, alertDecodeError}, err
+		// Decode error must be silently discarded
+		// [RFC6347 Section-4.1.2.7]
+		c.log.Debugf("defragment failed: %s", err)
+		return false, nil, nil
 	} else if isHandshake {
-		newHandshakeMessage := false
-		for out := c.fragmentBuffer.pop(); out != nil; out = c.fragmentBuffer.pop() {
+		for out, epoch := c.fragmentBuffer.pop(); out != nil; out, epoch = c.fragmentBuffer.pop() {
 			rawHandshake := &handshake{}
 			if err := rawHandshake.Unmarshal(out); err != nil {
-				return &alert{alertLevelFatal, alertDecodeError}, err
+				c.log.Debugf("%s: handshake parse failed: %s", srvCliStr(c.state.isClient), err)
+				continue
 			}
 
-			if c.handshakeCache.push(out, rawHandshake.handshakeHeader.messageSequence, rawHandshake.handshakeHeader.handshakeType, !c.state.isClient) {
-				newHandshakeMessage = true
-			}
-		}
-		if !newHandshakeMessage {
-			return nil, nil
+			_ = c.handshakeCache.push(out, epoch, rawHandshake.handshakeHeader.messageSequence, rawHandshake.handshakeHeader.handshakeType, !c.state.isClient)
 		}
 
-		c.lock.Lock()
-		defer c.lock.Unlock()
-		return c.handshakeMessageHandler(c)
+		return true, nil, nil
 	}
 
 	r := &recordLayer{}
 	if err := r.Unmarshal(buf); err != nil {
-		return &alert{alertLevelFatal, alertDecodeError}, err
+		return false, &alert{alertLevelFatal, alertDecodeError}, err
 	}
 
 	switch content := r.content.(type) {
 	case *alert:
-		c.log.Tracef("<- %s", content.String())
+		c.log.Tracef("%s: <- %s", srvCliStr(c.state.isClient), content.String())
+		var a *alert
 		if content.alertDescription == alertCloseNotify {
 			// Respond with a close_notify [RFC5246 Section 7.2.1]
-			_ = c.notify(alertLevelWarning, alertCloseNotify)
-			return nil, c.Close()
+			a = &alert{alertLevelWarning, alertCloseNotify}
 		}
-		return nil, fmt.Errorf("alert: %v", content)
+		return false, a, &errAlert{content}
 	case *changeCipherSpec:
-		c.log.Trace("<- ChangeCipherSpec")
+		if c.state.cipherSuite == nil || !c.state.cipherSuite.isInitialized() {
+			c.encryptedPackets = append(c.encryptedPackets, buf)
+			c.log.Debug("handleIncoming: CipherSuite not initialized, queuing packet")
+			return false, nil, nil
+		}
 
 		newRemoteEpoch := h.epoch + 1
-		if c.getRemoteEpoch() < newRemoteEpoch {
+		c.log.Tracef("%s: <- ChangeCipherSpec (epoch: %d)", srvCliStr(c.state.isClient), newRemoteEpoch)
+
+		if c.getRemoteEpoch()+1 == newRemoteEpoch {
 			c.setRemoteEpoch(newRemoteEpoch)
 		}
 	case *applicationData:
 		if h.epoch == 0 {
-			return &alert{alertLevelFatal, alertUnexpectedMessage}, fmt.Errorf("ApplicationData with epoch of 0")
+			return false, &alert{alertLevelFatal, alertUnexpectedMessage}, fmt.Errorf("ApplicationData with epoch of 0")
 		}
 
-		if err := c.handshakeErr.load(); err != nil {
-			return nil, err
-		}
-		if c.connectionClosed.Err() != nil {
-			return nil, io.EOF
-		}
+		c.closeLock.Lock()
+		connClosed := c.connectionClosed
```

Comment:
Can you make `c.connectionClosed` an atomic, or a function that auto locks/unlocks? Juggling mutexes has ended poorly for us a few times, so just want to get away from that.

Reply:
Read/write of `connectionClosedByUser` and `connectionClosed` must be synchronized at https://github.com/pion/dtls/pull/208/files#r388001174. Otherwise, it causes race condition.

Standard `net.Conn` implementations like `TCPConn` and `UDPConn` seems doesn't returning error from `Close()` at the first time of the connection is internally closed. This part is a logic for it.

Reply:
Adding `isConnectionClosed()` helper and some more comment.